### PR TITLE
Implement undocumented Twitch IRC tag "flags"

### PR DESCRIPTION
**Pull request by KararTY**
Based on https://github.com/Chatterino/chatterino2/issues/1772.

The `flags` irc tag was recently demystified, but is still undocumented, and there are still some questions regarding what the "score" actually is and the fact that categories and detection seems to not be the same day-after-day. Example\* is that the word `cunt` used to be *(July 1)* assigned 4 categories but testing it out today *(July 10th)* only returns 2 categories.

*Example for that phenomena:
```
[2020-07-1 21:54:48] #supinic supinic: $test cunt
[2020-07-1 21:54:48] #supinic supibot: 6-9:A.5/I.6/P.6/S.6
...
[2020-07-10 02:11:52] #supinic notkarar: $test cunt 󠀀
[2020-07-10 02:11:53] #supinic supibot: Automod score: Identity: 7; Profanity: 7 -- Raw: 6-9:I.7/P.7 󠀀
```
**Therefore the tests are as-is the day they were written, and the hardcoded values if they were to be compared to the values returned by Twitch, after some unknown time, will probably not be the same.**

There also used to exist some strange edge cases regarding links and drug substances where they wouldn't receive a category, but would still have `startIndex` & `endIndex`, I can't seem to repeat that behavior today (Related to above phenomena) but it's been observed:
```
[2020-07-1 22:19:41] #supinic dvh_: $test xanax
[2020-07-1 22:19:43] #supinic supibot: Automod score: -- Raw: 6-10:
```
I've covered that edge case with commit eaa7deb, for now, but I am sure there are better ways of covering it.

Looking forward for **feedback** on this pull request. @RAnders00

### CODE REVIEW

File: lib/message/flag.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * exclusive end index in the original message text.
+   * Note that we count unicode code points, not bytes with this.
+   * If you use this, make sure your code splits or indexes strings by their
+   * unicode code points, and not their bytes.
+   */
+  public endIndex: number;
+
+  /**
+   * The part of the original message string that was recognizes as flagged, e.g. "stfu".
```
Comment (RAnders00) [inline lib/message/flag.ts:22]:
```suggestion
   * The part of the original message string that was recognized as flagged, e.g. "stfu".
```

**Comment by KararTY** [comment]
Regexp: https://regexr.com/5892k

**Comment by RAnders00** [comment]
Sorry, I meant `USERNOTICE` messages, the ones that are sent during sub and resub events. Particularily the ones that actually have a message.

File: lib/message/twitch-types/usernotice.ts [hunk 2]
```diff
@@ -223,6 +224,9 @@ export class UsernoticeMessage extends ChannelIRCMessage {
   public readonly emotes: TwitchEmoteList;
   public readonly emotesRaw: string;
 
+  public readonly flags: TwitchFlagList | undefined;
+  public readonly flagsRaw: string | undefined;
```
Comment (RAnders00) [inline lib/message/twitch-types/usernotice.ts:228]: should add documentation to these entries

File: lib/message/twitch-types/privmsg.ts [hunk 3]
```diff
@@ -65,6 +66,9 @@ export class PrivmsgMessage extends ChannelIRCMessage
   public readonly emotes: TwitchEmoteList;
   public readonly emotesRaw: string;
 
+  public readonly flags: TwitchFlagList | undefined;
+  public readonly flagsRaw: string | undefined;
```
Comment (RAnders00) [inline lib/message/twitch-types/privmsg.ts:70]: should also add documentation here

**Comment by RAnders00** [comment]
Ignore the codeclimate bot's comments. I think the added code is fine.

File: lib/message/twitch-types/privmsg.ts [hunk 4]
```diff
@@ -66,7 +66,16 @@ export class PrivmsgMessage extends ChannelIRCMessage
   public readonly emotes: TwitchEmoteList;
   public readonly emotesRaw: string;
 
+  /**
+   * Can be an array of Twitch AutoMod flagged words, for use in moderation and/or filtering purposes.
+   *
+   * **Note:** This is an undocumented Twitch IRC feature and may change at any time, use at your own risk.
```
Comment (RAnders00) [inline lib/message/twitch-types/privmsg.ts:72]:
```suggestion
   * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
   * attributes which when missing or malformed will fail the message parsing. However since this attribute is
   * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
   * While it will be a major version release if this attribute changes format in dank-twitch-irc, using this is still
   * at your own risk since it may suddenly contain unexpected data or turn `undefined` one day as
   * Twitch changes something. In short: **Use at your own risk** and make sure your
   * implementation can handle the case where this is `undefined`.
```

File: lib/message/twitch-types/privmsg.ts [hunk 5]
```diff
@@ -66,7 +66,16 @@ export class PrivmsgMessage extends ChannelIRCMessage
...
+   */
   public readonly flags: TwitchFlagList | undefined;
+
+  /**
+   * Twitch AutoMod raw flags string.
```
Comment (RAnders00) [inline lib/message/twitch-types/privmsg.ts:77]:
```suggestion
   * Twitch AutoMod raw flags string.
   *
   * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
   * attributes which when missing or malformed will fail the message parsing. However since this attribute is
   * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
   * In short, ensure your implementation can handle the case where this is `undefined` or is in
   * a format you don't expect.
```

File: lib/message/twitch-types/usernotice.ts [hunk 6]
```diff
@@ -224,7 +224,16 @@ export class UsernoticeMessage extends ChannelIRCMessage {
   public readonly emotes: TwitchEmoteList;
   public readonly emotesRaw: string;
 
+  /**
+   * Can be an array of Twitch AutoMod flagged words, for use in moderation and/or filtering purposes.
+   *
+   * **Note:** This is an undocumented Twitch IRC feature and may change at any time, use at your own risk.
```
Comment (RAnders00) [inline lib/message/twitch-types/usernotice.ts:230]:
```suggestion
   * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
   * attributes which when missing or malformed will fail the message parsing. However since this attribute is
   * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
   * While it will be a major version release if this attribute changes format in dank-twitch-irc, using this is still
   * at your own risk since it may suddenly contain unexpected data or turn `undefined` one day as
   * Twitch changes something. In short: **Use at your own risk** and make sure your
   * implementation can handle the case where this is `undefined`.
```

File: lib/message/twitch-types/usernotice.ts [hunk 7]
```diff
@@ -224,7 +224,16 @@ export class UsernoticeMessage extends ChannelIRCMessage {
...
+   */
   public readonly flags: TwitchFlagList | undefined;
+
+  /**
+   * Twitch AutoMod raw flags string.
```
Comment (RAnders00) [inline lib/message/twitch-types/usernotice.ts:235]:
```suggestion
   * Twitch AutoMod raw flags string.
   *
   * If the `flags` tag is missing or of a unparseable format, this will be `undefined`. This is unlike most other
   * attributes which when missing or malformed will fail the message parsing. However since this attribute is
   * completely undocumented we cannot rely on the `flags` tag being stable, so this soft fallback is used instead.
   * In short, ensure your implementation can handle the case where this is `undefined` or is in
   * a format you don't expect.
```
